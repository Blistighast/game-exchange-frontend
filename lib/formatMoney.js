export default function formatMoney(amount = 0) {
  // set to 0 for it doesnt break if nothing is passed
  const options = {
    style: 'currency',
    currency: 'USD',
    minimumFractionDigits: 2,
  };

  // check if its a clean dollar amount, if so, remove the decimals
  if (amount % 100 === 0) {
    options.minimumFractionDigits = 0;
  }

  // configures numbers into money decimals, $ sign and commas based on given location
  const formatter = Intl.NumberFormat('en-US', options); // can write the options here directly as well

  return formatter.format(amount / 100);
}
