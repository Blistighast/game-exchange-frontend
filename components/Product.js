import Link from 'next/link';
import ItemStyles from './styles/ItemStyles';
import Title from './styles/Title';
import PriceTag from './styles/PriceTag';
import formatMoney from '../lib/formatMoney';
import DeleteProduct from './DeleteProduct';
import AddToCart from './AddToCart';
import NotSignedIn from './NotSignedIn';

export default function Product({ product }) {
  return (
    <div>
      <ItemStyles>
        <img
          src={product?.photo?.image?.publicUrlTransformed} // ? checks if it exists before continuing
          alt={product.name}
        />
        <Title>
          <div>
            <Link href={`/product/${product.id}`}>{product.name}</Link>
          </div>
        </Title>
        <PriceTag>{formatMoney(product.price)}</PriceTag>
        <p>{product.description.length > 200 
          ? product.description.substring(0, 200) + '...' 
          : product.description}</p>
        <div className="buttonList">
          <NotSignedIn>
            <Link
              href={{
                pathname: 'update',
                query: {
                  id: product.id,
                },
              }}
            >
            {/* <Link href={`/update/${product.id}`}> */}
              Edit ðŸ›  {/* window + . gives you emoji options */}
            </Link>
            <AddToCart id={product.id} />
            <DeleteProduct id={product.id}>Delete</DeleteProduct>
          </NotSignedIn>
        </div>
      </ItemStyles>
    </div>
  );
}
