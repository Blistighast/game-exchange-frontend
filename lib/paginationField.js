import gql from 'graphql-tag';
import { PAGINATION_QUERY } from '../components/Pagination';

export default function paginationField() {
  return {
    keyArgs: false, // tells apollo we will take care of everything, dont use default
    read(existing = [], { args, cache }) {
      // args are the first and skip values, cache is existing cache
      const { skip, first } = args;

      // read the number of items on the cache
      const data = cache.readQuery({ query: PAGINATION_QUERY });
      const count = data?._allProductsMeta?.count;
      const page = skip / first + 1;
      const pages = Math.ceil(count / first);

      // check if we have existing items
      const items = existing.slice(skip, skip + first).filter((x) => x);
      // if
      // there are items
      // AND there arent enough items to satisfy how many were requested
      // AND we're on the last page
      // then just send it
      if (items.length && items.length !== first && page === pages) {
        return items;
      }
      if (items.length !== first) {
        // we dont have any items, we must go to network to fetch
        return false;
      }
      // if there are items, just return them from the cache, dont need to go to network
      if (items.length) {
        return items;
      }

      return false; // fallback to network if both if statements dont work
      // first thing apollo does is askes the read function for those items
      // we can either do 1 of 2 things
      // first: can return the items because they are already in the cache
      // two: return false from here, (make a new network request)
    },

    merge(existing, incoming, { args }) {
      const { skip, first } = args;
      // this runs when apollo client comes back from network with out products
      const merged = existing ? existing.slice(0) : []; // if there is cache makes a copy
      for (let i = skip; i < skip + incoming.length; i++) {
        merged[i] = incoming[i - skip];
      }
      // finally we return the merged items from the cache
      return merged;
    },
  };
}
