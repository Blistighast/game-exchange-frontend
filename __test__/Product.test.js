import { MockedProvider } from "@apollo/react-testing";
import { render, screen } from "@testing-library/react";
import { fakeItem } from "../lib/testUtils";
import Product from "../components/Product";

const product = fakeItem();

describe("<Product/>", () => {
  it("renders out the title and price tag", () => {
    const { container, debug } = render(
      // need to use the mock provider because of apollo, pretends component is wrapped in the apollo provider
      <MockedProvider>
        <Product product={product} />
      </MockedProvider>
    );
    const priceTag = screen.getByText('$50')
    const link = container.querySelector('a')
    expect(priceTag).toBeInTheDocument();
    expect(link).toHaveAttribute('href', '/product/abc123'); // can use direct string
    expect(link).toHaveTextContent(product.name); // or use the variable from the props
  });

  it('renders and matches the snapshot', () => {
    const { container, debug } = render(
      <MockedProvider>
        <Product product={product} />
      </MockedProvider>
    );
    expect(container).toMatchSnapshot();
  })
});
