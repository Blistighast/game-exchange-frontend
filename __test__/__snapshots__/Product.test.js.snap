// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Product/> renders and matches the snapshot 1`] = `
<div>
  <div>
    <div
      class="ItemStyles-sc-1luezei-0 jSObhF"
    >
      <img
        alt="dogs are best"
        src="dog.jpg"
      />
      <h3
        class="Title-sc-1b2xigg-0 eaTBBs"
      >
        <div>
          <a
            href="/product/abc123"
          >
            dogs are best
          </a>
        </div>
      </h3>
      <span
        class="PriceTag-sc-14ws024-0 dySomE"
      >
        $50
      </span>
      <p>
        dogs
      </p>
      <div
        class="buttonList"
      />
    </div>
  </div>
</div>
`;
